### ISSUE TITLE
DWARF gets incorrect DW_AT_linkage_name for generics

### ISSUE THREAD

**Issue by tromey** [rust-lang/rust#32925]
Consider a source file like:

```
fn identity<A>(x: A) -> A { return x; }

fn main() {
    println!("{}", identity(5i32));
    println!("{}", identity(5u64));
}
```

When compiled with `rustc -g`, the generated DWARF has incorrect values for `DW_AT_linkage_name`.  These should agree with `nm` but do not:

```
bapiya. readelf -wi ba |grep identity
    <19c>   DW_AT_linkage_name: (indirect string, offset: 0x3c8): _ZN2ba13identity<i32>E
    <1a0>   DW_AT_name        : (indirect string, offset: 0x3df): identity<i32>
    <1d1>   DW_AT_linkage_name: (indirect string, offset: 0x41d): _ZN2ba13identity<u64>E
    <1d5>   DW_AT_name        : (indirect string, offset: 0x434): identity<u64>
bapiya. nm ba | grep identity
0000000000005160 t _ZN8identity21h16171468230716265285E
00000000000051a0 t _ZN8identity21h18186590313173030232E
```

**Comment by michaelwoerister** [comment]
Yes, that's a known issue. In former times, it was the only way to get GDB to print nice, fully-qualified function names. We might want to revisit that though.

**Comment by tromey** [comment]
Nightly is emitting values that are even less valid, like:

```
    <db>   DW_AT_linkage_name: (indirect string, offset: 0x422): _ZN7methods8{{impl}}3newE
```

**Comment by steveklabnik** [comment]
Triage: not sure if anything has changed here, but I don't think so.

**Comment by tmiasko** [comment]
This was fixed at some point, probably by #46899:

```console
$ rustc ba.rs -g
$ readelf -wi ba |grep identity
    <386>   DW_AT_linkage_name: (indirect string, offset: 0x42d): _ZN2ba8identity17h4032c34ace3f366cE
    <38a>   DW_AT_name        : (indirect string, offset: 0x451): identity<u64>
    <3bb>   DW_AT_linkage_name: (indirect string, offset: 0x45f): _ZN2ba8identity17h97d8ddefaee8e693E
    <3bf>   DW_AT_name        : (indirect string, offset: 0x483): identity<i32>
$ nm ba | grep identity
0000000000007c20 t _ZN2ba8identity17h4032c34ace3f366cE
0000000000007c30 t _ZN2ba8identity17h97d8ddefaee8e693E
```
